### ISSUE TITLE
v22 crash when the message Upgraded from v21 to v22 occurs

### ISSUE THREAD

**Issue by Pekkape** [llde/xOBSE#87]
I tried 2 different savegames and one in a very tiny cell. First save the game frooze with the message:
OBSE is now upgraded from v21 to v22

The second save file, the one in the small tiny cell did CTD. I installed 21.8 and everything works great. 

The question now is if I am the only one with this issue and if I am not, what I can do to help solving this problem as I really looked forward to 22.0 and really want to be able to run 22.x +

**Comment by Pekkape** [comment]
Here is a list of my plugins:
AddActorValues.dll
AddActorValues_CS.dll
Blockhead.dll
Blockhead.ini
ConScribe.dll
ConScribe.ini
Construction Set Extender.dll
Construction Set Extender.ini
CSE
Elys_USV.dll
Elys_USV.lip
Elys_USV.mp3
Elys_USV.txt
EngineBugFixes.dll
EngineBugFixes.ini
FasterSleepWait.dll
FasterSleepWait.ini
FastExit2.dll
MenuQue
MenuQue.dll
MenuQue.ini
MessageLogger.dll
MessageLogger.ini
obse_jail_fix.dll
OBSE_Kyoma_MenuQue.dll
obse_training_fix.dll
PekkasOBSEfile.txt
SkyBSA.dll
SkyBSA.log
SkyBSA.log0
sr_Oblivion_Stutter_Remover.dll
sr_Oblivion_Stutter_Remover.ini

The sr_Oblivion_Stutter_Remover.ini is from https://www.nexusmods.com/oblivion/mods/47439 and I set the heap size to 1000 to see if I can fast travel with the map, which has stopped working for me a week ago but one of these plugins might cause my CTD.

**Comment by Pekkape** [comment]
Here is a textfile, of my esp's
[Pekkasesplist.txt](https://github.com/llde/xOBSE/files/6215532/Pekkasesplist.txt)

**Comment by llde** [comment]
In this specific case I'm more interested in the backtrace instead of load order.
I have all the plugins you have + other ones. 
OSR heap should rarely be set to 1000, but I doubt it has any relevance

If you can get a debugger working and provide me the stack trace using the symbols from the PDB I'm good. Otherwise you may try the debug DLL  from AppVeyor

**Comment by Pekkape** [comment]
I did install the appveyor-server and have it running and what else do I need to do?  I did run the game again with v22 but the timestamps on the PDB files did not change, which they should have if a change was made to them? Do I need to set a specific flag when running xOBSE to get the debug info? How do I proceed exactly as I never done this before?

**Comment by llde** [comment]
No you misunderstood me. 
I configured AppVeyor  to build a debug/release pair, that can be downlaoded by users.
In this case I aksed you to try to download  the debig DLL from the AppVeyor  link from the xOBSE page (build:success button in the readme)

**Comment by Pekkape** [comment]
I looked around at https://github.com/llde/xOBSE for something you tried to make me find, well except for https://ci.appveyor.com/project/llde/xOBSE/branch/master/job/y1bv7ai4q6yr4y2g which does not make me wiser, so I have to give up. I do not follow what to do really. Could you provide a link and instruction?

**Comment by llde** [comment]
To grab the dll from AppVeyor you have to go to the Artifacts Tab, it has a different name as it's not deployed. The second link is almost there.
Of course PDB are also different between builds.
But again anything mostly useless if you can't get  a  backtrace of the crash.

Do you have CobbCrashLogger?

**Comment by Pekkape** [comment]
I will install CobbCrashLogger now and we see what info it might be able to provide. Sounds like a good idea really. I will take a break now and I will test in an hour.

**Comment by Pekkape** [comment]
AHHH. Oki, I found it, the debug version, finally. It wasn't easy nor obvious really to find the right spot where you had hidden it, but I was on the right track earlier at least, very close and still so far away... ;)

 Well I start with the original release first with Crashlogger and then I try the debug version, together with crashlogger again I guess. Here is that first Crash log
[CobbCrashLogger.log](https://github.com/llde/xOBSE/files/6216256/CobbCrashLogger.log)

**Comment by Pekkape** [comment]
The debug files have different names then the original files have. Do I keep those names you gave them or do I rename them to the Deployment name's?

**Comment by llde** [comment]
Should be the original name

**Comment by Pekkape** [comment]
Before I ran it, I changed the heap size back to 768 as I did test 1000 as someone else managed to get fast travel to work with that size and that size only in an older thread at Nexus. Do you recommend any other value?

Well here is the OBSE.log when running the 5 debug version files. I used the save in that small cell that CTD, not freezing the game as other save files I have do with v22.
[obse.log](https://github.com/llde/xOBSE/files/6216342/obse.log)

**Comment by llde** [comment]
Apparently this bug is the same Laulajatar reported, caused by a regression in hotkeys functions, or at least a bug with a very similar sintomatology. It's crashing in BaseExtraList::GetByType

**Comment by Pekkape** [comment]
I see. But that's 6 months ago and I should than had the same issue I guess, which I did not?

**Comment by llde** [comment]
@Pekkape are you on the xOBSE discord?

**Comment by llde** [comment]
@Pekkape can you check if your issues persist with this DLL? 
https://ci.appveyor.com/project/llde/xOBSE/branch/master/job/glawcblsrcrsc228/artifacts

**Comment by Pekkape** [comment]
Nope, I am not in the xOBSE Discord as I rarely run Discord but I do when needed. 
I did tried the new version but the game freeze at the same spot still.

**Comment by llde** [comment]
@Pekkape I may need a more precise level of information, can you activate obse minidumps and send one to me?
it's the bCreateCrashDump=0 option in obse.ini in Data\OBSE
Put bCreateCrashDump=1 to activate

**Comment by Pekkape** [comment]
No action. It freeze at the same spot, as soon as the message shows up that OBSE is now upgrade from v21 to v22. As it freeze, it might not be considered as a crash? I do need to logout and login to be able to close the game down. 
bCreateCrashDump is now set to 1

This is the latest file created in the Oblivion directory...  ;)
[obse.log](https://github.com/llde/xOBSE/files/6254420/obse.log)

There is a cell I could place my char at that makes the game CTD instead of just freeze, if needed. I do not have that save file saved where I where able to CTD while testing, if it now matters that is.

**Comment by llde** [comment]
@Pekkape 

> There is a cell I could place my char at that makes the game CTD instead of just freeze, if needed. I do not have that save file saved where I where able to CTD while testing, if it now matters that is.
If you can please try to trigger the crash. A CTD is more valuable then an hang in debuggability.
It may also be valuable if you can do a bisection on the modlist to understand which mod cause this issue

**Comment by Pekkape** [comment]
Bisection?  I did include my modlist earlier, and I have not changed it, if that is what you mean?

**Comment by llde** [comment]
For bisection I mean a process of disabling/enabling mods to check which one cause a certain problem.
It's called bisection becouse it's similar to a binary search  pattern.
IT consist in disabling half mod in the load order, try, if still the bug happen, disable another half, again, if the bug doesn't happen renable half of the mods just disabled, check and so on, until you understand what mod cause the issue.

Of course, you can avoid disabling good know mods as the uop.
THe bisection mechanism is efficient as  it require at most logn n passes, where n is the number of esp/esm plugins installed.
If you suspect a mod to cause the issue (in this case a mod that  tackle with inventory objects and references),  then you may limit trying to disable the suspected mod and try.

**Comment by Pekkape** [comment]
Oki. I could do that but another way is that I upload the save file as well, if you wanna make tests with it, which might be best really.

**Comment by llde** [comment]
You can do it, but the save may be useful only if I know what mod cause the issue in the firstplace.

**Comment by Pekkape** [comment]
I can do it without the Wrye patch as well, as remaking it takes extremely long time - To long time for my patience for sure.

**Comment by llde** [comment]
keep the bashed patch always disabled during the bisection.

**Comment by Pekkape** [comment]
Oki. I might done 20 tests and got it to work when I unchecked Auto Repair https://www.nexusmods.com/oblivion/mods/38582 with now freeze nor CTD

I am rebuilding the patch now, enabling all mods I wanna use, except that one and I will soon make a final test

**Comment by Pekkape** [comment]
When I activated all my mods again, except Auto Repair, and patched it, it freeze at the same point.

I did see in the OBSE.log warnings about 3 mods that have loads of strings and those strings are meant to be temporary from both the efshop and my mod, the Hobbit Home:

Loading strings
  WARNING: substantial numbers of string variables exist for the following files (may indicate savegame bloat):
    Ultimate Leveling.esp (644 strings)
    efShopMod.esp (66909 strings)
    hobbitHome.esp (39844 strings)

I did peek in the obse save file that is now 7k big, and the majority of junk saved in it, is temporary strings, well strings that are meant to be temporary, stuff I do printc on to see if my dynamic sorter is doing what it is supposed to do, which it does.

**Comment by Pekkape** [comment]
Now I did uncheck my mod, hobbitHome.esp and it is taken down for the moment (I cannot say why but it is not because of a technical issue) but I can share it with you from my Google Drive. It is big though 262Mb and it do have 150-151 scripts in it but I bet it is all those junk temp strings that is produced and saved that might be the problem.

I did test first to run efshop without hobbithome and it did work, even with all those junk strings loaded and it did work just fine with no freeze. Then I unchecked efshop and ran my mod, and it frooze. So 2 mods is causing the freeze, HobbitHome and Auto Repair.

**Comment by llde** [comment]
efShopMod.esp (66909 strings)
hobbitHome.esp (39844 strings)

This is a lot of junk in the save.
string_var aren't cleaned automatically, they must be destroyed with sv_destruct. Once a string is in the cosave, they are in the cosave until the mod is removed. (I'm thinking to sync with latest xNVSE changes that resolve this issue in the future)

thanks I will try to reproduce with autorepair.

**Comment by Pekkape** [comment]
So in the future we do not need to use sv_destruct or what you meant is, if I start to destroy my temporary strings with sv_destruct, they might be able to get deleted from the save file as well in a future release? So this is already solved in xNVSE?

**Comment by llde** [comment]
Yes the string bloat was already solved in xNVSE.
I think it's better to manually destruct the string every time, but I still plan to backport the xNVSE changes into xOBSE.
Note that message appearing (" WARNING: substantial numbers of string variables exist for the following files (may indicate savegame bloat):") may also be caused by completly legitimate ways of using strings, and don't actually indicate bloat.

**Comment by Pekkape** [comment]
Oki. I do try to clean the save file, to see if it is possible or not, as I will not mod more today as the OBSE save do seems like a pure textfile really - am I right? I will soon figure it out though and I did a backup anyway. Tomorrow I will look through all my scripts, and make them more proper and really delete a lot of unnecessary debug stuff I have added and add sv_destruct to every string I have, well after they have been used that is...  ;)   

I could just empty the house but it is loaded with stuff, so that will not be an easy task. Empty the efshop and restart that one is easier for sure.

**Comment by Pekkape** [comment]
I did managed to delete 500k until I got tired of in it the OBSE save, loaded the game and saw a black screen...  Well it was a bad idea...

**Comment by llde** [comment]
@Pekkape Can you try the debug DLL from here: 
https://ci.appveyor.com/project/llde/xOBSE/build/job/l6b5iyycjryp1xkj/artifacts
And if it hang or crash send me the obse.log

**Comment by Pekkape** [comment]
Tomorrow if it is okay with you?

Well guaranteeing stuff is like that is never good to do...  ;)   Lets hope for the best anyway...   ;)  My game is getting very unstable right now so I should do something about those 2 mods right away.

**Comment by llde** [comment]
A well, tomorrow is fine.
BTW also AutoRepair leak strings

**Comment by Pekkape** [comment]
Ahhh   :)   Good to know as it does feel we got somewhere really as it seems the handling of strings is the main problem here, right?

**Comment by Pekkape** [comment]
I was thinking to clean up all those strings first before I make any more tests really and it might not be tomorrow but within a 2-3 days I guess. I decided to clean out my house, putting the stuff in temp chests elsewhere as I do not want to delete anything I collected and I am a real junk collector for sure and that is what my mod is all about anyway, making your own museum. Well I will make a save file without those 3 mods activated but before that, I will add sv_destruct to all my scripts using strings, and lots of them do.

**Comment by llde** [comment]
umh. knowing why it hang with the new version but not with the old, even if the save may be dirty, maybe ther eis something that can be made more robust, to prevent this from happening.

**Comment by Pekkape** [comment]
I have cleaned my mod now but some strings are impossible to destruct as I do not use them as temp strings

I took a look at Auto Repair and he doesn't use strings at all but arrays

Efshop is filled with strings in every script and he use strings for loads of different stuff, stuff that could be shorts instead really, but I only took a quick peek as maybe they cannot just be shorts as that is a guess and they can't be destructed, not as they are used now but when I looked at the names and what they seems to be used for, it would have been easiest to use shorts as flags as I got the impression they are flags, nothing else and if it is so, then the use of strings is not really the proper way to go.

So now I need to move everything to a temp mod and if all goes well, I might be able to make tests tomorrow already. So much stuff...  I plan to dig my own grave as my prime save file is very close to 40Mb so something will soon break for sure, no matter what.

**Comment by llde** [comment]
Auto Repair construct strings every time to generate the message to print to the ui.
they are generated instead of being reused every time the repair script is picked.

The ones that are persistent across the game are fines even if not destructed.

**Comment by Pekkape** [comment]
> umh. knowing why it hang with the new version but not with the old, even if the save may be dirty, maybe ther eis something that can be made more robust, to prevent this from happening.

It depends really, depending on what you did edit in v22. ;)  I did start to read about the changes but there was also so many.

About Auto repair: Well i couldn't find any strings defined in any of those scripts, but I might missed something as I did sort them in CSE bit it doesn't matter as you did find it and that is more important for sure for the bigger picture, than what I find. I was more curious to see if I could make a quick fix or not really... ;)

It took almost 8 hours to clean my mod and also move all stuff to a temp mod but the OBSE save file looks so tiny and neat now and got it down 11x from 7Mb to 660kb and I can even use fast travel now as well as that stopped to work 2 weeks ago so all this was needed for sure as I was extremely close to a complete melt down. Making a save file also goes 3 times faster.

**Comment by llde** [comment]
Well cosave save and load aren't optimized the best.
I do plan to optimize them in the future, after the more fundamental changes.

From arrepairscript

	if brokenHammers > 0
		if brokenHammers == 1
			set strBrokenHammers to sv_Construct "A repair hammer broke,"
		else
			set strBrokenHammers to sv_Construct "%.0f repair hammers broke," brokenHammers
		endif

**Comment by Pekkape** [comment]
Ahh, so that is possible without defining or declaring the strings first?  I would per auto define them and I did only look in the header, well before the begin in each script so I would had written:

string_var strBrokenHammers 

begin blablabla

if brokenHammers > 0
	if brokenHammers == 1
		set strBrokenHammers to sv_Construct "A repair hammer broke,"
	else
		set strBrokenHammers to sv_Construct "%.0f repair hammers broke," brokenHammers
	endif

...but maybe that is not needed then?

**Comment by Pekkape** [comment]
I found it and it will be easy to fix at a later point for sure for myself, but I keep the mod untouched as long as we have this issue open or are in need of it unchanged so to say. The mod is 8 years old so it is unlikely I will be able to reach the author...

**Comment by Pekkape** [comment]
Test run. First test I did was with hobbithome.esp activated and now it works. I stayed still in game for a couple of minutes as the FPS was 2-3 with this debug version.

2nd test was with AutoRepair.esp activated, hobbitHome.esp turned off and followed did happen.

1. I got repair hammers
2. The message came up: Updating OBSE from v21 to v22
3. 1-3 seconds later it did repair my items
4. FREEZE

I do get the feeling it freeze when it tries to use that string undeclared, just a hunch and wild guess though. I will declare it in the beginning of that script and run that mod again and see what happens.

**Comment by Pekkape** [comment]
I edited it, changed the definition to:
`string_var strBrokenHammers`
but it still freezes, so it was not as easy as I thought.
[obse.log](https://github.com/llde/xOBSE/files/6257787/obse.log)

**Comment by llde** [comment]
Can you try reproducing this bug with the last xOBSE and the last CobbCrashLogger?

**Comment by Pekkape** [comment]
Sorry for a late reply Idle. I will do it during the week. At the moment I got a bad tooth ache so I will wait a couple of days until I try it, just so you know I have not ignored you...  ;)

I did restart a new game as my save file got so damn huge so maybe I cannot reproduce the bug but we find out sooner or later. My save file got up to 60Mb and then I uninstalled some mods and it went to 55Mb and I thought...  Screw it. What will happen at 60Mb is crashes every time I try to move to another cell, either through doors, coc or moveto. Moving to a map marker was out of question even at the 55Mb size...

**Comment by Pekkape** [comment]
The game did freeze and here is the messagelog made with the latest CobbCrashLogger and also the obse.log if you need it.
[MessageLog.txt](https://github.com/llde/xOBSE/files/6587145/MessageLog.txt)
[obse.log](https://github.com/llde/xOBSE/files/6587146/obse.log)

**Comment by llde** [comment]
that MessageLog is messed up a lot.

**Comment by Pekkape** [comment]
Yes...   ;)  By the way, it did freeze after the message popped up - Updating OBSE from 21 to 22 and the message got away and freeze.

**Comment by llde** [comment]
@Pekkape  Hi. Can you check if using this DLL it still hangs?
https://ci.appveyor.com/api/buildjobs/y8k4g2efotml0451/artifacts/obse%2Fobse%2FRelease%201_2_0_416%2Fobse.dll

**Comment by Pekkape** [comment]
I did a quick login and tested v22.6.1 and it works. Even Denock arrow with COBL works and i did had problems with that after v21.5 and I did test some other v22 and i will now make a longer test with this new version. I just wanted to be sure it works with COBL as I will not play without COBL. That is out of the question  :D  

I also got rid of all String bloats. My current game seems clean as it can be (I hope). My obse save file is steady in size and saving is fast. I have looked in all my scripts, and I edited ef Shop for personal use, destruct every damn string in all those strings scripts as they all use strings and the Author has not been available for 2 years now. Many things are broken in my personal copy but the main stuff in the mod works and that is enough for me _(I cheat, fixing a full working shop with some edits in some scripts)_. I will make a new trading mod from scratch soon, without strings, or as few as possible at least, with proper sv_destruct in each script if I do. We see...  If it is needed, it is. We started a WIP for it 2 weeks ago at Nexus mnod section, debating the basics of what we wanna have included. It will work similar to Skyrim Merchant Stall...  Immersive as it can be  :D    

Well why do I mention it and babble about trading? Because ef Shop mod did cause this problem in the first place + my own strings in my own scripts in my own WIP.   All clean now.
